packages/plans-grid-next/src/components/comparison-grid/hooks/useVisibleGridPlans.ts

```diff
@@ -0,0 +1,76 @@
+import {
+	PlanSlug,
+	URL_FRIENDLY_TERMS_MAPPING,
+	UrlFriendlyTermType,
+	findFirstSimilarPlanKey,
+	isFreePlan,
+} from '@automattic/calypso-products';
+import { WpcomPlansUI } from '@automattic/data-stores';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect } from 'react';
+import { usePlansGridContext } from '../../../grid-context';
+import { GridPlan } from '../../../types';
+
+/**
+ *
+ * When intervalType change trickles down to the ComparisonGrid as a prop (intervalType),
+ * we must detect this prop change and transform the current visible plan order into a set of plans
+ * that align with the currently selected plan interval without causing glitches or flashes.
+ *
+ * This hook serves as the solution. The hook accomplishes two main tasks:
+ * - Resolves the correct visible plan order, providing it immediately in the current render cycle to prevent glitches.
+ * - Updates the WpcomPlansUI store with the most up-to-date plan order as required.
+ *
+ * ******
+ * @param {UrlFriendlyTermType} intervalType - Selected plan interval
+ * @returns The list of gridPlans sorted by the accurate compact view order.
+ */
+
+function useVisibleGridPlans( intervalType?: UrlFriendlyTermType ): GridPlan[] {
+	const { gridPlansIndex } = usePlansGridContext();
+	const { setComparisonGridVisibleGridPlans } = useDispatch( WpcomPlansUI.store );
+	const visibleGridPlansStoreState = useSelect(
+		( select ) => select( WpcomPlansUI.store ).getComparisonGridVisibleGridPlans(),
+		[]
+	);
+
+	let refreshedVisibleGridPlans: PlanSlug[] = visibleGridPlansStoreState;
+	let isCompactViewStale = false;
+
+	/**
+	 * We do not check for staleness if the interval type is not provided
+	 */
+	if ( visibleGridPlansStoreState.length > 0 && intervalType ) {
+		/**
+		 * If at least one planSlug in the visibleGridPlansStoreState state is not present in the current displayedPlanSlugs,
+		 * it indicates that the visibleGridPlansStoreState is now stale most probably due to a change in the interval.
+		 * This is because plan slugs are unique per term.
+		 */
+		isCompactViewStale = visibleGridPlansStoreState.some( ( plan ) => ! gridPlansIndex[ plan ] );
+
+		if ( isCompactViewStale ) {
+			refreshedVisibleGridPlans = visibleGridPlansStoreState.map( ( previousSlug ) => {
+				if ( isFreePlan( previousSlug ) ) {
+					return previousSlug;
+				}
+				return findFirstSimilarPlanKey( previousSlug, {
```

Comment:
I wonder if it's a question about what's simpler:

```
getPlanClass( A ) === getPlanClass( B )
```

and

```
comparePlans( A, B, [ 'type' ] ) === true
```

Is there an actual need to generalize into a wider abstraction? Meaning, are there current uses outside of checking `type`? and assuming this is a utility worth having, then we need to refactor other cases in the codebase, otherwise this becomes just additional code/liability.

I suggest we stick with `getPlanClass` and not introduce the abstraction here. You can do that in a separate PR and  also refactor all other instances of the first case above (there are a few). We can also focus on reviewing that utility as a separate thing.